modules/misc/nix-remote-build.nix

```diff
@@ -0,0 +1,219 @@
+# Manages the remote build configuration
+{ config, lib, ... }:
+
+let
+  inherit (lib)
+    optional
+    concatMapStrings
+    concatStringsSep
+    mkIf
+    mkOption
+    optionalString
+    types
+    versionAtLeast
+    getVersion
+    ;
+
+  cfg = config.nix;
+
+  nixPackage = cfg.package;
+
+  isNixAtLeast = versionAtLeast (getVersion nixPackage);
+
+  buildMachinesText = concatMapStrings (
+    machine:
+    (concatStringsSep " " (
+      [
+        "${optionalString (machine.protocol != null) "${machine.protocol}://"}${
+          optionalString (machine.sshUser != null) "${machine.sshUser}@"
+        }${machine.hostName}"
+        (concatStringsSep "," machine.systems)
+        (if machine.sshKey != null then machine.sshKey else "-")
+        (toString machine.maxJobs)
+        (toString machine.speedFactor)
+        (
+          let
+            res = (machine.supportedFeatures ++ machine.mandatoryFeatures);
+          in
+          if (res == [ ]) then "-" else (concatStringsSep "," res)
+        )
+        (
+          let
+            res = machine.mandatoryFeatures;
+          in
+          if (res == [ ]) then "-" else (concatStringsSep "," machine.mandatoryFeatures)
+        )
+      ]
+      ++ optional (isNixAtLeast "2.4pre") (
+        if machine.publicHostKey != null then machine.publicHostKey else "-"
+      )
+    ))
+    + "\n"
+  ) cfg.buildMachines;
+
+in
+{
+  options = {
+    nix = {
+      buildMachines = mkOption {
+        type = types.listOf (
+          types.submodule {
```

Comment:
question: Would there be any benefit to freeform support? Or has this interface been pretty stable for a long time?